### PR TITLE
fabtests/fi_inj_complete: Add  FI_INJECT to fabtests and change fi_inj_complete to a generic flag test

### PR DESCRIPTION
**Pull request by Juee14Desai**
There are 2 commits in this pull request:

- Remove -R option from fi_inj_complete because FI_INJECT and FI_INJECT_COMPLETE are send only calls.
- Make arguments in inj_complete.c align with the other tests. Use of -m (size) and -O (iterations) in inj_complete.c was not needed and hence, this has been changed to generic -S and -I respectively. send_msg argument has been changed to -N since now -S is used for size.
- Add FI_INJECT test and validation support to fabtests. New argument -A can be used for testing FI_INJECT and FI_INJECT_COMPLETE.
- Change test fi_inj_complete to fi_msg_flags and filename inj_complete.c to msg_flags.c. The newly renamed file is more generic and new msg flags can be added later on for testing.
- Add new FI_INJECT tests to runfabtests.sh script.

### CODE REVIEW

File: fabtests/functional/inj_complete.c [hunk 1]
```diff
@@ -188,11 +183,9 @@ int main(int argc, char **argv)
 			break;
 		case '?':
 		case 'h':
-			ft_usage(argv[0], "FI_Inject_Completion Functional Test");
-			FT_PRINT_OPTS_USAGE("-m <size>", "size of Injection message");
-			FT_PRINT_OPTS_USAGE("-S", "enable testing with fi_sendmsg");
+			ft_csusage(argv[0], "FI_Inject_Completion Functional Test");
+			FT_PRINT_OPTS_USAGE("-N", "enable testing with fi_sendmsg");
 			FT_PRINT_OPTS_USAGE("-R", "enable testing with fi_recvmsg");
```
Comment (shefty) [inline fabtests/functional/inj_complete.c:207]: Out of curiosity, why is there an fi_recvmsg test option?  That seems unrelated to the purpose of the test.

File: fabtests/functional/inj_complete.c [hunk 2]
```diff
@@ -146,7 +155,7 @@ static int run_test(void)
 				return ret;
 		}
 	}
-	fprintf(stdout, "GOOD: Completed FI_INJECT_COMPLETIONS Testing\n");
+	fprintf(stdout, "GOOD: Completed %s Testing\n", fi_tostr(&flag, FI_TYPE_OP_FLAGS));
```
Comment (shefty) [inline fabtests/functional/inj_complete.c:158]: Please check line lengths here and above and below, and limit to around 80 characters.

File: fabtests/functional/inj_complete.c [hunk 3]
```diff
@@ -181,12 +190,23 @@ int main(int argc, char **argv)
...
 			opts.options |= FT_OPT_VERIFY_DATA;
 			break;
+		case 'A':
+			if (!strncasecmp("inj_complete", optarg, 12)) {
+				flag = FI_INJECT_COMPLETE;
+			} else if (!strncasecmp("inject", optarg, 6)) {
+				flag = FI_INJECT;
+			} else {
+				printf ("Unsupported flag\n");
+				return EXIT_FAILURE;
```
Comment (shefty) [inline fabtests/functional/inj_complete.c:200]: I would consider renaming the test to maybe fi_inject_test.  The current name aligns with testing for inject_complete semantics.

I would go even further and just have the test verify both options and avoid needing to pass in the parameters.  We should look at doing this for other tests, so that running a test can validate more of the API without needing to re-run the test with a different set of options.

File: fabtests/scripts/runfabtests.sh [hunk 4]
```diff
@@ -130,12 +130,10 @@ functional_tests=(
...
-	"fi_inj_complete -e msg"
-	"fi_inj_complete -e rdm"
-	"fi_inj_complete -e dgram"
-	"fi_inj_complete -e msg -SR"
-	"fi_inj_complete -e rdm -SR"
-	"fi_inj_complete -e dgram -SR"
+	"fi_msg_flags -e rdm -A inject -v"
+	"fi_msg_flags -e rdm -NR -A inject -v"
+	"fi_msg_flags -e rdm -A inj_complete -v"
+	"fi_msg_flags -e rdm -NR -A inj_complete -v"
```
Comment (shefty) [inline fabtests/scripts/runfabtests.sh:136]: If the test only supports RDM endpoints, we should have the test use that by default and remove needing to pass that in.

**Review by shefty — COMMENTED** [review]
Comments were made reviewing each path individually, so the comment about renaming the test only partially applies.  I kept the comment, because I would consider keeping inject as part of the name (fi_msg_inject), since that's all it tests.  

We should have a general discussion on this, but the route I would look toward moving to is to reduce the number of command line options each of these test takes as input.  Instead, have the test iterate over all of its available options so that we get the broadest coverage possible that a test provides.  For this specific PR, that would mean having fi_msg_flags automatically check both FI_INJECT and FI_INJECT_COMPLETE without requiring the user to run both.

File: fabtests/functional/inj_complete.c (hunk 1 above)
Comment (Juee14Desai) [reply fabtests/functional/inj_complete.c]: fi_recvmsg was not needed and I removed that and related code in a separate commit.

File: fabtests/functional/inj_complete.c (hunk 2 above)
Comment (Juee14Desai) [reply fabtests/functional/inj_complete.c]: Thanks for pointing it out. Addressed in latest changes.

File: fabtests/functional/inj_complete.c (hunk 3 above)
Comment (Juee14Desai) [reply fabtests/functional/inj_complete.c:183]: - I renamed it to fi_msg_inject and also renamed the file to fi_msg_inject.c as I wanted to keep the names similar. Is that ok or would you want me to rename the test to fi_inject_test? 

- I was trying to avoid the need to pass a parameter for FI_INJECT and FI_INJECT_COMPLETE but I found out that non sendmsg path that requires the EP to be configured with OP flags so it would be configured with both and so iterating through all flags would require the test to re-initialize to test each case. So I went with the current way.

File: fabtests/scripts/runfabtests.sh (hunk 4 above)
Comment (Juee14Desai) [reply fabtests/scripts/runfabtests.sh]: Yes I agree. RDM is made default in the latest changes.

**Comment by shefty** [comment]
CI failures:
```
10:44:09  - name:   fi_msg_inject -NR -A inject -v -p "tcp"
10:44:09    result: Fail
10:44:09    time:   1
10:44:09    server_cmd: /home/build/ofi-Install/libfabric/ofi_libfabric/PR-7298/6/dbg/bin/fi_msg_inject -NR -A inject -v -p "tcp"  -s n7
10:44:09    server_stdout: |
10:44:09      /home/build/ofi-Install/libfabric/ofi_libfabric/PR-7298/6/dbg/bin/fi_msg_inject: invalid option -- 'R'
```

File: fabtests/scripts/runfabtests.sh [hunk 5]
```diff
@@ -130,12 +130,10 @@ functional_tests=(
...
-	"fi_inj_complete -e msg"
-	"fi_inj_complete -e rdm"
-	"fi_inj_complete -e dgram"
-	"fi_inj_complete -e msg -SR"
-	"fi_inj_complete -e rdm -SR"
-	"fi_inj_complete -e dgram -SR"
+	"fi_msg_inject -A inject -v"
+	"fi_msg_inject -NR -A inject -v"
+	"fi_msg_inject -A inj_complete -v"
+	"fi_msg_inject -NR -A inj_complete -v"
```
Comment (shefty) [inline fabtests/scripts/runfabtests.sh:136]: Not sure what the 'R' is doing here after the N for the above 2 calls.

Comment (Juee14Desai) [reply fabtests/scripts/runfabtests.sh]: I missed the R here. All checks are passing now.

File: fabtests/scripts/runfabtests.sh [hunk 6]
```diff
@@ -130,12 +130,10 @@ functional_tests=(
...
-	"fi_inj_complete -e dgram"
-	"fi_inj_complete -e msg -S"
-	"fi_inj_complete -e rdm -S"
-	"fi_inj_complete -e dgram -S"
+	"fi_msg_inject -A inject -v"
+	"fi_msg_inject -N -A inject -v"
+	"fi_msg_inject -A inj_complete -v"
+	"fi_msg_inject -N -A inj_complete -v"
 	"fi_bw -e rdm -v -T 1"
 	"fi_bw -e rdm -v -T 1 -U"
```
Comment (aingerson) [inline fabtests/scripts/runfabtests.sh:138]: Could you move this change to the beginning-ish of the list of commits (specifically to at least before you rename the test? This patch changes a lot of lines that were previously modified so putting it in the correct order will reduce the number of changes.
Also please move the -S -> -N into the patch that changes the flag

File: fabtests/functional/inj_complete.c [hunk 7]
```diff
@@ -149,37 +160,41 @@ int main(int argc, char **argv)
...
+			}
+			break;
 		case '?':
 		case 'h':
-			ft_usage(argv[0], "FI_Inject_Completion Functional Test");
-			FT_PRINT_OPTS_USAGE("-m <size>", "size of Injection message");
-			FT_PRINT_OPTS_USAGE("-S", "enable testing with fi_sendmsg");
-			FT_PRINT_OPTS_USAGE("-O", "enable testing injection overrun");
+			ft_csusage(argv[0], "FI_Inject_Completion Functional Test");
```
Comment (aingerson) [inline fabtests/functional/inj_complete.c:188]: this usage should get edited in your rename patch

File: fabtests/functional/inj_complete.c [hunk 8]
```diff
@@ -149,37 +160,41 @@ int main(int argc, char **argv)
...
+			FT_PRINT_OPTS_USAGE("-N", "enable testing with fi_sendmsg");
```
Comment (aingerson) [inline fabtests/functional/inj_complete.c:189]: This patch should also change the -S in runfabtests to be -N since you change the argument in this patch

File: fabtests/scripts/runfabtests.sh [hunk 9]
```diff
@@ -130,12 +130,10 @@ functional_tests=(
...
-	"fi_inj_complete -e msg"
-	"fi_inj_complete -e rdm"
-	"fi_inj_complete -e dgram"
-	"fi_inj_complete -e msg -S"
-	"fi_inj_complete -e rdm -S"
-	"fi_inj_complete -e dgram -S"
+	"fi_msg_inject -A inject -v"
+	"fi_msg_inject -N -A inject -v"
+	"fi_msg_inject -A inj_complete -v"
+	"fi_msg_inject -N -A inj_complete -v"
```
Comment (aingerson) [inline fabtests/scripts/runfabtests.sh:136]: You change runfabtests to use the new name but your rename patch comes after. Move the runfabtests rename to your rename patch

**Review by aingerson — COMMENTED** [review]
The order and contents of your commits is still off. Let's simplify and have 2 commits:
- one to refactor/update the existing test (remove -R, change -S to -N and use -I and -S, and make this a rdm test)
- one to add FI_INJECT support, rename the test, and add the new testing in runfabtests
Applied in that order. That should simplify the patch modification issues.

**Comment by shefty** [comment]
@aingerson - Do you have any more comments on these changes?  I think the issues you raised were addressed.

**Review by aingerson — APPROVED** [review]
@shefty Nope, looks good to me

**Comment by Juee14Desai** [comment]
@shefty Thank you for merging these changes.
